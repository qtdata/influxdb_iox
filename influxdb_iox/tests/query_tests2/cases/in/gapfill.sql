-- Gap-filling tests
-- IOX_SETUP: OneMeasurementTwoSeries

-- Input data
-- region=a 2000-05-05T12:20:00Z
-- region=a 2000-05-05T12:40:00Z
-- region=b 2000-05-05T12:31:00Z
-- region=b 2000-05-05T12:39:00Z

-- IOX_COMPARE: uuid
EXPLAIN SELECT
  date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute,
  count(cpu.user)
from cpu
where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z'
group by minute;

-- Gap filling with no other group keys
SELECT
  date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute,
  count(cpu.user)
from cpu
where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z'
group by minute;

-- Gap filling with no other group keys and no aggregates
SELECT
  date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute
from cpu
where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z'
group by minute;

-- gap filling with a group key
SELECT
  date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute,
  region,
  count(cpu.user)
from cpu
where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z'
group by minute, region;

-- gap filling with an origin argument that is not the epoch
SELECT
  date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:07Z') as minute,
  region,
  count(cpu.user)
from cpu
where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z'
group by minute, region;

