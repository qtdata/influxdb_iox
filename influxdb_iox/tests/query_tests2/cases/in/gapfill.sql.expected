-- Test Setup: OneMeasurementTwoSeries
-- SQL: EXPLAIN SELECT date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute, count(cpu.user) from cpu where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z' group by minute;
-- Results After Normalizing UUIDs
----------
| plan_type    | plan    |
----------
| logical_plan    | Projection: date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z")) AS minute, COUNT(cpu.user)    |
|    |   GapFill: groupBy=[[date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))]], aggr=[[COUNT(cpu.user)]], time_column=date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z")), stride=IntervalDayTime("600000"), range=Included(TimestampNanosecond(957528000000000000, None))..Included(TimestampNanosecond(957531540000000000, None))    |
|    |     Aggregate: groupBy=[[datebin(IntervalDayTime("600000"), cpu.time, TimestampNanosecond(0, None)) AS date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))]], aggr=[[COUNT(cpu.user)]]    |
|    |       Filter: cpu.time >= TimestampNanosecond(957528000000000000, None) AND cpu.time <= TimestampNanosecond(957531540000000000, None)    |
|    |         TableScan: cpu projection=[time, user], partial_filters=[cpu.time >= TimestampNanosecond(957528000000000000, None), cpu.time <= TimestampNanosecond(957531540000000000, None)]    |
| physical_plan    | ProjectionExec: expr=[date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))@0 as minute, COUNT(cpu.user)@1 as COUNT(cpu.user)]    |
|    |   GapFillExec: group_expr=[date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))@0], aggr_expr=[COUNT(cpu.user)@1], stride=600000, time_range=Included("957528000000000000")..Included("957531540000000000")    |
|    |     SortExec: expr=[date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))@0 ASC]    |
|    |       CoalescePartitionsExec    |
|    |         AggregateExec: mode=FinalPartitioned, gby=[date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))@0 as date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))], aggr=[COUNT(cpu.user)]    |
|    |           CoalesceBatchesExec: target_batch_size=8192    |
|    |             RepartitionExec: partitioning=Hash([Column { name: "date_bin_gapfill(IntervalDayTime(\"600000\"),cpu.time,Utf8(\"1970-01-01T00:00:00Z\"))", index: 0 }], 4), input_partitions=4    |
|    |               AggregateExec: mode=Partial, gby=[datebin(600000, time@0, 0) as date_bin_gapfill(IntervalDayTime("600000"),cpu.time,Utf8("1970-01-01T00:00:00Z"))], aggr=[COUNT(cpu.user)]    |
|    |                 RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1    |
|    |                   CoalesceBatchesExec: target_batch_size=8192    |
|    |                     FilterExec: time@0 >= 957528000000000000 AND time@0 <= 957531540000000000    |
|    |                       ParquetExec: limit=None, partitions={1 group: [[1/1/1/1/00000000-0000-0000-0000-000000000000.parquet]]}, predicate=time >= TimestampNanosecond(957528000000000000, None) AND time <= TimestampNanosecond(957531540000000000, None), pruning_predicate=time_max@0 >= 957528000000000000 AND time_min@1 <= 957531540000000000, projection=[time, user]    |
|    |    |
----------
-- SQL: SELECT date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute, count(cpu.user) from cpu where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z' group by minute;
+----------------------+-----------------+
| minute               | COUNT(cpu.user) |
+----------------------+-----------------+
| 2000-05-05T12:00:00Z |                 |
| 2000-05-05T12:10:00Z |                 |
| 2000-05-05T12:20:00Z | 1               |
| 2000-05-05T12:30:00Z | 2               |
| 2000-05-05T12:40:00Z | 1               |
| 2000-05-05T12:50:00Z |                 |
+----------------------+-----------------+
-- SQL: SELECT date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute from cpu where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z' group by minute;
+----------------------+
| minute               |
+----------------------+
| 2000-05-05T12:00:00Z |
| 2000-05-05T12:10:00Z |
| 2000-05-05T12:20:00Z |
| 2000-05-05T12:30:00Z |
| 2000-05-05T12:40:00Z |
| 2000-05-05T12:50:00Z |
+----------------------+
-- SQL: SELECT date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute, region, count(cpu.user) from cpu where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z' group by minute, region;
+----------------------+--------+-----------------+
| minute               | region | COUNT(cpu.user) |
+----------------------+--------+-----------------+
| 2000-05-05T12:00:00Z | a      |                 |
| 2000-05-05T12:10:00Z | a      |                 |
| 2000-05-05T12:20:00Z | a      | 1               |
| 2000-05-05T12:30:00Z | a      |                 |
| 2000-05-05T12:40:00Z | a      | 1               |
| 2000-05-05T12:50:00Z | a      |                 |
| 2000-05-05T12:00:00Z | b      |                 |
| 2000-05-05T12:10:00Z | b      |                 |
| 2000-05-05T12:20:00Z | b      |                 |
| 2000-05-05T12:30:00Z | b      | 2               |
| 2000-05-05T12:40:00Z | b      |                 |
| 2000-05-05T12:50:00Z | b      |                 |
+----------------------+--------+-----------------+
-- SQL: SELECT date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:07Z') as minute, region, count(cpu.user) from cpu where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z' group by minute, region;
+----------------------+--------+-----------------+
| minute               | region | COUNT(cpu.user) |
+----------------------+--------+-----------------+
| 2000-05-05T11:50:07Z | a      |                 |
| 2000-05-05T12:00:07Z | a      |                 |
| 2000-05-05T12:10:07Z | a      | 1               |
| 2000-05-05T12:20:07Z | a      |                 |
| 2000-05-05T12:30:07Z | a      | 1               |
| 2000-05-05T12:40:07Z | a      |                 |
| 2000-05-05T12:50:07Z | a      |                 |
| 2000-05-05T11:50:07Z | b      |                 |
| 2000-05-05T12:00:07Z | b      |                 |
| 2000-05-05T12:10:07Z | b      |                 |
| 2000-05-05T12:20:07Z | b      |                 |
| 2000-05-05T12:30:07Z | b      | 2               |
| 2000-05-05T12:40:07Z | b      |                 |
| 2000-05-05T12:50:07Z | b      |                 |
+----------------------+--------+-----------------+